### PR TITLE
Secret Link as a subservice

### PR DESCRIPTION
**Pull request by ines-cruz**
closes #https://github.com/inveniosoftware/invenio-rdm-records/issues/530

### CODE REVIEW

File: invenio_rdm_records/ext.py [hunk 1]
```diff
@@ -22,8 +22,9 @@
     RDMRecordFilesResourceConfig, RDMRecordResource, RDMRecordResourceConfig
 from .secret_links import LinkNeed, SecretLink
 from .services import RDMFileDraftServiceConfig, RDMFileRecordServiceConfig, \
-    RDMRecordService, RDMRecordServiceConfig
+    RDMRecordService, RDMRecordServiceConfig, SecretLinkServiceConfig
 from .services.schemas.metadata_extensions import MetadataExtensions
+from .services.secret_links.service import SecretLinkService
```
Comment (ppanero) [inline invenio_rdm_records/ext.py:27]: When refactoring one of the main points is to try not to have to change imports. This is to avoid "breaking compatibilty". In this case you can avoid having this import (and leaving it as it was) by "exporting" the class in the `__init__.py`, note that you can do that successively (until you export it in `.services.__init__.py`).

File: invenio_rdm_records/ext.py [hunk 2]
```diff
@@ -106,8 +107,8 @@ def init_services(self, app):
             self._filter_record_service_config(app, RDMRecordServiceConfig),
             files_service=FileService(RDMFileRecordServiceConfig),
             draft_files_service=FileService(RDMFileDraftServiceConfig),
+            secret_links_service=SecretLinkService(SecretLinkServiceConfig)
```
Comment (ppanero) [inline invenio_rdm_records/ext.py:110]: 💯

File: invenio_rdm_records/services/__init__.py [hunk 3]
```diff
@@ -8,7 +8,7 @@
 """High-level API for wokring with RDM records, files, pids and search."""
 
 from .config import RDMFileDraftServiceConfig, RDMFileRecordServiceConfig, \
-    RDMRecordServiceConfig
+    RDMRecordServiceConfig, SecretLinkServiceConfig
```
Comment (ppanero) [inline invenio_rdm_records/services/__init__.py:11]: For my comment on the `ext.py`... you are actually exporting it, so there is no need to change the import (in `ext.py`)

File: invenio_rdm_records/services/config.py [hunk 4]
```diff
@@ -225,3 +225,50 @@ class RDMFileDraftServiceConfig(FileServiceConfig):
         "content": FileLink("{+api}/records/{id}/draft/files/{key}/content"),
         "commit": FileLink("{+api}/records/{id}/draft/files/{key}/commit"),
     }
+
+
+class SecretLinkServiceConfig(RecordServiceConfig):
```
Comment (ppanero) [inline invenio_rdm_records/services/config.py:230]: moderate: The config is 100% the same than the `RecordServiceConfig`? If so there is no need to create it... You can instantiate the subservice by passing the `RecordServiceConfig`.

File: invenio_rdm_records/services/secret_links/service.py [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+import arrow
+from flask_babelex import lazy_gettext as _
+from invenio_db import db
+from invenio_drafts_resources.services.records import RecordService
+from invenio_records_resources.services.records.schema import \
+    ServiceSchemaWrapper
+from marshmallow.exceptions import ValidationError
+
+
+class SecretLinkService(RecordService):
```
Comment (ppanero) [inline invenio_rdm_records/services/secret_links/service.py:21]: inheritance/design: `SecretLinkService` inherits from `RecordsService` which after a few parental relations end up in [here](https://github.com/inveniosoftware/invenio-records-resources/blob/master/invenio_records_resources/services/records/service.py#L23). 

Before, the functions were named `{create, read, etc.}_secret_link` to avoid overwriting a record one. But now that it is a subservice... if they follow the same interface (they should) they can be renamed to `read`, `create` etc.

File: invenio_rdm_records/services/secret_links/service.py [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+    """RDM Secret Link service."""
+
+    def __init__(self, config, secret_links_service=None):
+        """Constructor for Secret Link Service."""
+        super().__init__(config)
+        self._secret_links = secret_links_service
```
Comment (ppanero) [inline invenio_rdm_records/services/secret_links/service.py:27]: major: I think there is a conceptual error...  The *secret_links* service should be "independent" like e.g. [`FileService`](https://github.com/inveniosoftware/invenio-records-resources/blob/master/invenio_records_resources/services/files/service.py#L22) and inherit from `Service`. Then, it would be `RDMRecordsService` who would have the `secret_links` attribute and property (i.e. `RDMRecordsService` would do something similar to [`RecordsService`](https://github.com/inveniosoftware/invenio-drafts-resources/blob/master/invenio_drafts_resources/services/records/service.py#L31)...

All this said, maybe I'm missing a reason why it has to be like these 😅

**Review by ppanero — COMMENTED** [review]
Looks geeting in shape :)

TL;DR:
- Try to see if it can be done without changing imports
- Update copyright to YYYY-2021 to the files you touch if they did not have 2021 yet (e.g. 2020-2021)
- major: service inheritance instantiation might not be correct... crosscheck against draft-resources/records-resources links added in comments.

File: invenio_rdm_records/ext.py [hunk 7]
```diff
@@ -24,6 +24,7 @@
 from .services import RDMFileDraftServiceConfig, RDMFileRecordServiceConfig, \
     RDMRecordService, RDMRecordServiceConfig
 from .services.schemas.metadata_extensions import MetadataExtensions
+from .services.secret_links import SecretLinkService
```
Comment (ppanero) [inline invenio_rdm_records/ext.py:27]: moderate: it should be possible to import from "services", by exposing it in the corresponding `__init__.py`.

```suggestion
from .services import RDMFileDraftServiceConfig, RDMFileRecordServiceConfig, \
    RDMRecordService, RDMRecordServiceConfig, SecretLinkService
from .services.schemas.metadata_extensions import MetadataExtensions
```

File: invenio_rdm_records/resources/resources.py [hunk 8]
```diff
@@ -1,6 +1,6 @@
 # -*- coding: utf-8 -*-
 #
-# Copyright (C) 2020-2021 CERN.
+# Copyright (C) 2021 CERN.
```
Comment (ppanero) [inline invenio_rdm_records/resources/resources.py:3]: major: should be a range, we should only add... never remove
```suggestion
# Copyright (C) 2020-2021 CERN.
```

File: invenio_rdm_records/resources/resources.py [hunk 9]
```diff
@@ -149,7 +149,7 @@ def delete(self):
     @response_handler(many=True)
     def search(self):
```
Comment (ppanero) [inline invenio_rdm_records/resources/resources.py:150]: minor: I'm doubting if we should actually call this `read_all`... because in the end we are not passing any "search query"

File: invenio_rdm_records/services/secret_links/service.py [hunk 10]
```diff
@@ -0,0 +1,287 @@
...
+        # Fetching
+        links = parent.access.links.resolve_all()
+        return self.link_result_list(
+            service=self,
+            identity=identity,
+            results=links,
+            links_config=links_config,
+        )
+
+    def read_link(
```
Comment (ppanero) [inline invenio_rdm_records/services/secret_links/service.py:172]: minor: just `read`?

File: invenio_rdm_records/services/services.py [hunk 11]
```diff
@@ -1,6 +1,6 @@
 # -*- coding: utf-8 -*-
 #
-# Copyright (C) 2020 CERN.
+# Copyright (C) 2021 CERN.
```
Comment (ppanero) [inline invenio_rdm_records/services/services.py:3]: should not remove, just update ranges.

File: tests/secret_links/test_sharing.py [hunk 12]
```diff
@@ -1,5 +1,6 @@
 # -*- coding: utf-8 -*-
 #
+# Copyright (C) 2021 CERN.
```
Comment (ppanero) [inline tests/secret_links/test_sharing.py:3]: info: in this case is correct because is the first time we (CERN) touch it.

File: invenio_rdm_records/services/__init__.py (hunk 3 above)
Comment (ppanero) [reply invenio_rdm_records/services/__init__.py]: still present

File: invenio_rdm_records/services/secret_links/service.py [hunk 13]
```diff
@@ -0,0 +1,287 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2021 CERN.
```
Comment (lnielsen) [inline invenio_rdm_records/services/secret_links/service.py:3]: You will need to keep the copyright header from the original file here (it's important because it assigns ownership over the source code, and in this specific case it was done by TU Wien :-))

```
# Copyright (C) 2020-2021 CERN.
# Copyright (C) 2020-2021 Northwestern University.
# Copyright (C) 2021 TU Wien.
```

**Review by ppanero — APPROVED** [review]
